src/core/ddsc/src/dds_qos.c

```diff
@@ -335,6 +335,151 @@ void dds_qset_ignorelocal (dds_qos_t * __restrict qos, dds_ignorelocal_kind_t ig
+    return false;
+  for (uint32_t i = 0; !found && i < qos->property.binary_value.n; i++)
+  {
+    if (strcmp (qos->property.binary_value.props[i].name, name) == 0)
+    {
+      *index = i;
+      found = true;
+    }
+  }
+  return found;
+}
+
+void dds_qset_prop (dds_qos_t * __restrict qos, const char * name, const char * value)
+{
+  uint32_t i;
+  if (qos == NULL || name == NULL || value == NULL)
+    return;
+
+  dds_qprop_init (qos);
+  if (dds_qprop_get_index (qos, name, &i))
+  {
+    dds_free (qos->property.value.props[i].value);
+    qos->property.value.props[i].value = dds_string_dup (value);
+  }
+  else
+  {
+    qos->property.value.props = dds_realloc (qos->property.value.props,
+      (qos->property.value.n + 1) * sizeof (*qos->property.value.props));
+    qos->property.value.props[qos->property.value.n].propagate = 0;
+    qos->property.value.props[qos->property.value.n].name = dds_string_dup (name);
+    qos->property.value.props[qos->property.value.n].value = dds_string_dup (value);
+    qos->property.value.n++;
+  }
+}
+
+void dds_qunset_prop (dds_qos_t * __restrict qos, const char * name)
+{
+  uint32_t i;
+  if (qos == NULL || !(qos->present & QP_PROPERTY_LIST) || qos->property.value.n == 0 || name == NULL)
+    return;
+
+  if (dds_qprop_get_index (qos, name, &i))
+  {
+    dds_free (qos->property.value.props[i].name);
+    dds_free (qos->property.value.props[i].value);
+    if (qos->property.value.n > 1)
+    {
+      if (i < (qos->property.value.n - 1))
+        memmove (qos->property.value.props + i, qos->property.value.props + i + 1,
+          (qos->property.value.n - i - 1) * sizeof (*qos->property.value.props));
+
+      qos->property.value.props = dds_realloc (qos->property.value.props,
+        (qos->property.value.n - 1) * sizeof (*qos->property.value.props));
+    }
+    else
+    {
+      dds_free (qos->property.value.props);
+      qos->property.value.props = NULL;
+    }
+    qos->property.value.n--;
+  }
+}
+
+void dds_qset_bprop (dds_qos_t * __restrict qos, const char * name, const void * value, const size_t sz)
+{
+  uint32_t i;
+  if (qos == NULL || name == NULL || (value == NULL && sz > 0))
+    return;
+
+  dds_qprop_init (qos);
+  if (dds_qbprop_get_index (qos, name, &i))
+  {
+    dds_qos_data_copy_in (&qos->property.binary_value.props[i].value, value, sz, true);
+  }
+  else
+  {
+    qos->property.binary_value.props = dds_realloc (qos->property.binary_value.props,
+      (qos->property.binary_value.n + 1) * sizeof (*qos->property.binary_value.props));
+    qos->property.binary_value.props[qos->property.binary_value.n].propagate = 0;
+    qos->property.binary_value.props[qos->property.binary_value.n].name = dds_string_dup (name);
+    dds_qos_data_copy_in (&qos->property.binary_value.props[qos->property.binary_value.n].value, value, sz, false);
+    qos->property.binary_value.n++;
+  }
+}
+
+void dds_qunset_bprop (dds_qos_t * __restrict qos, const char * name)
```

Comment:
Similar to the ``get_index`` functions, I would combine the two ``unsets``. I'd probably do the same for the part of the ``set`` functions that extends the array, but I'm not entirely certain that would be worth the bother.

Reply:
See above. I didn't refactor the `set` functions, because I think it's not worth the effort here